doc/langref.html.in

```diff
@@ -6899,8 +6899,255 @@ fn foo() i32 {
 
       {#header_open|Result Location Semantics#}
       <p>
-      <a href="https://github.com/ziglang/zig/issues/2809">TODO add documentation for this</a>
+      During compilation, every Zig expression and sub-expression is assigned optional result location
+      information. This information dictates what type the expression should have (its result type), and
+      where the resulting value should be placed in memory (its result location). This is not an
+      implementation detail: this logic is codified into the Zig language specification. This system is
+      collectively referred to as "Result Location Semantics".
       </p>
+      {#header_open|Result Types#}
+      <p>
+      Result types are propagated recursively through expressions where possible. For instance, if the
+      expression {#syntax#}&e{#endsyntax#} has result type {#syntax#}*u32{#endsyntax#}, then
+      {#syntax#}e{#endsyntax#} is given a result type of {#syntax#}u32{#endsyntax#}, allowing the
+      language to perform this coercion before taking a reference.
+      </p>
+      <p>
+      The result type mechanism is utilized by casting builtins such as {#syntax#}@intCast{#endsyntax#}.
+      Rather than taking as an argument the type to cast to, these builtins use their result type to
+      determine this information. The result type is often known from context; where it is not, the
+      {#syntax#}@as{#endsyntax#} builtin can be used to explicitly provide a result type.
+      </p>
+      <p>
+      We can break down the result types for each component of a simple expression as follows:
+      </p>
+      {#code_begin|test|result_type_propagation#}
+test "result type propagates through struct initializer" {
+    const S = struct { x: u32 };
+    const val: u64 = 123;
+    const s: S = .{ .x = @intCast(val) };
+    // .{ .x = @intCast(val) }   has result type `S` due to the type annotation
```

Comment:
The phrase "has result type" should probably make it clearer if it "has" the type because of RLS inference, or "has" the type because of source annotations.  Perhaps something like:

   // .{ .x = @intCast(val) }  is given a result type `S` inferred from the explicit type annotation on `s`

Reply:
I don't understand what you mean here. To say an expression "has result type" `T` means, precisely and unambiguously, that this expression was assigned this type due to the mechanics of RLS. There is no practical or semantic difference between a directly-annotated result type or an inferred result type for a more nested expression.